tooltester/src/main/java/io/dockstore/tooltester/S3Client.java

```
@@ -0,0 +1,116 @@
+            String[] encodedToolIdArray = Arrays.copyOf(split, split.length - 1);
+            return String.join("/", encodedToolIdArray);
+        } else {
+            return toolId;
+        }
+    }
+
+    /**
+     * Create s3 object with metadata and upload it to s3
+     * @param toolId    The TRS ToolId
+     * @param versionName   The TRS ToolVersion name
+     * @param testFilePath  The file that was tested (test1.json, Dockerfile)
+     * @param runner    The runner used to test the file (cwltool, cromwell)
+     * @param logContent    The contents of the log file from ToolTester
+     * @param startTime     The start time (seconds since epoch) when the file was tested
+     */
+    public void createObject(String toolId, String versionName, String testFilePath, String runner, String logContent, String startTime) {
+        try {
+            String key = generateKey(toolId, versionName, testFilePath, runner, startTime);
+            byte[] contentAsBytes = logContent.getBytes(StandardCharsets.UTF_8);
+            ByteArrayInputStream contentsAsStream = new ByteArrayInputStream(contentAsBytes);
+            ObjectMetadata metadata = new ObjectMetadata();
+            metadata.setContentType(MediaType.TEXT_PLAIN);
+            metadata.addUserMetadata("tool_id", toolId);
+            metadata.addUserMetadata("version_name", versionName);
+            metadata.addUserMetadata("test_file_path", testFilePath.replaceFirst("^Build ", "").replaceFirst("^Test ", ""));
+            metadata.addUserMetadata("runner", runner);
+            metadata.setContentLength(contentAsBytes.length);
+            PutObjectRequest request = new PutObjectRequest(bucketName, key, contentsAsStream, metadata);
+            s3.putObject(request);
+        } catch (UnsupportedEncodingException e) {
+            System.out.println("Could not generate S3 URL: " + e.getMessage());
```

1. Use Logger
2. This is going to silently fail. Is that OK?

---

2. Outputting a message is hardly considered silent (since this is a CLI) :smile:

---

I didn't know if this was being invoked from a web service endpoint when I ran across the code.

If it is CLI, then do you want to do a System.exit(<non zero value>).? Dockstore CLI has an `exceptionMessage` function for this sort of thing...

---

I prefer it to continue generating csv report even if it's not complete.